cdap-security/src/test/java/co/cask/cdap/security/authorization/InMemoryAuthorizer.java

```diff
@@ -33,25 +33,35 @@
 import com.google.common.collect.Sets;
 import com.google.common.collect.Table;
 
-import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Properties;
 import java.util.Set;
-import javax.annotation.concurrent.NotThreadSafe;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 
 /**
  * In-memory implementation of {@link Authorizer}.
  */
-@NotThreadSafe
 public class InMemoryAuthorizer extends AbstractAuthorizer {
 
   private final Table<EntityId, Principal, Set<Action>> table = HashBasedTable.create();
```

Comment:
If you use a `ConcurrentMap<Principal, ConcurrentMap<Entity, Set<Action>>>`, then you don't need the lock